File: solutions/2025-12-31/solution.java
// Minimum Window Substring
// Difficulty: Hard
// URL: https://leetcode.com/problems/minimum-window-substring/

public String minWindow(String s, String t) {
  if (s == null || t == null || s.length() == 0 || t.length() == 0) {
    return "";
  }
  int m = s.length(), n = t.length();
  int[] tCount = new int[128];
  for (char c : t.toCharArray()) {
    tCount[c]++;
  }
  int required = 0;
  for (int count : tCount) {
    if (count > 0) {
      required++;
    }
  }
  int left = 0;
  int minLen = Integer.MAX_VALUE;
  String minWindow = "";
  int formed = 0;
  int[] windowCounts = new int[128];
  for (int right = 0; right < m; right++) {
    char c = s.charAt(right);
    windowCounts[c]++;
    if (tCount[c] > 0 && windowCounts[c] == tCount[c]) {
      formed++;
    }
    while (left <= right && formed == required) {
      c = s.charAt(left);
      if (right - left + 1 < minLen) {
        minLen = right - left + 1;
        minWindow = s.substring(left, right + 1);
      }
      windowCounts[c]--;
      if (tCount[c] > 0 && windowCounts[c] < tCount[c]) {
        formed--;
      }
      left++;
    }
  }
  return minWindow;
}

File: solutions/2025-12-31/solution.py
# Minimum Window Substring
# Difficulty: Hard
# URL: https://leetcode.com/problems/minimum-window-substring/

def minWindow(s, t):
  if not s or not t:
    return ""
  m, n = len(s), len(t)
  t_count = {}
  for char in t:
    t_count[char] = t_count.get(char, 0) + 1
  required = len(t_count)
  left = 0
  min_len = float('inf')
  min_window = ""
  formed = 0
  window_counts = {}
  for right in range(m):
    character = s[right]
    window_counts[character] = window_counts.get(character, 0) + 1
    if character in t_count and window_counts[character] == t_count[character]:
      formed += 1
    while left <= right and formed == required:
      character = s[left]
      if right - left + 1 < min_len:
        min_len = right - left + 1
        min_window = s[left:right + 1]
      window_counts[character] -= 1
      if character in t_count and window_counts[character] < t_count[character]:
        formed -= 1
      left += 1
  return min_window